File: app/src/main/java/com/Cherry/sme/tts/MainActivity.java
package com.cherry.sme.tts;

import android.app.Activity;
import android.app.ProgressDialog;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.PowerManager;
import android.preference.PreferenceManager;
import android.provider.Settings;
import android.view.View;
import android.view.WindowManager;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;
import java.util.ArrayList;
import java.util.List;

public class MainActivity extends Activity {

    private SharedPreferences prefs;
    private ArrayList<String> engineNames = new ArrayList<>();
    private ArrayList<String> enginePackages = new ArrayList<>();
    private ProgressDialog progressDialog;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        prefs = PreferenceManager.getDefaultSharedPreferences(this);

        showScanDialog();

        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                loadInstalledEngines();

                setupEngineUI(R.id.spinnerShan, "pref_engine_shan", "com.espeak.ng");
                setupEngineUI(R.id.spinnerBurmese, "pref_engine_myanmar", "org.saomaicenter.myanmartts");
                setupEngineUI(R.id.spinnerEnglish, "pref_engine_english", "com.google.android.tts");

                if (progressDialog != null && progressDialog.isShowing()) {
                    progressDialog.dismiss();
                }

                getWindow().clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
            }
        }, 1500);

        setupDonation(R.id.btnKpay, "09750091817", "KBZ Pay Number Copied");
        setupDonation(R.id.btnWave, "09750091817", "Wave Pay Number Copied");

        setupOpenSystemSettings(R.id.spinnerShan);
        setupOpenSystemSettings(R.id.spinnerBurmese);
        setupOpenSystemSettings(R.id.spinnerEnglish);

        setupBatteryOptimization();
        setupAutoStart();
        updateBatteryStatus();
    }

    private void setupBatteryOptimization() {
        View btn = findViewById(R.id.btnBatteryOptimize);
        btn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                requestIgnoreBatteryOptimization();
            }
        });
    }

    private void requestIgnoreBatteryOptimization() {
        try {
            PowerManager pm = (PowerManager) getSystemService(Context.POWER_SERVICE);
            if (pm != null && !pm.isIgnoringBatteryOptimizations(getPackageName())) {
                Intent intent = new Intent(Settings.ACTION_REQUEST_IGNORE_BATTERY_OPTIMIZATIONS);
                intent.setData(Uri.parse("package:" + getPackageName()));
                startActivity(intent);
                LogCollector.addLog("BATTERY", "Requested ignore battery optimization");
            } else {
                Toast.makeText(this, "Battery optimization already disabled", Toast.LENGTH_SHORT).show();
                LogCollector.addLog("BATTERY", "Already ignoring battery optimization");
                openBatterySettings();
            }
        } catch (Exception e) {
            LogCollector.addError("BATTERY", "Failed to request battery optimization", e);
            openBatterySettings();
        }
    }

    private void openBatterySettings() {
        try {
            Intent intent = new Intent(Settings.ACTION_IGNORE_BATTERY_OPTIMIZATION_SETTINGS);
            startActivity(intent);
        } catch (Exception e) {
            try {
                Intent intent = new Intent(Settings.ACTION_BATTERY_SAVER_SETTINGS);
                startActivity(intent);
            } catch (Exception e2) {
                Toast.makeText(this, "Please manually disable battery optimization in Settings", Toast.LENGTH_LONG).show();
            }
        }
    }

    private void setupAutoStart() {
        View btn = findViewById(R.id.btnAutoStart);
        btn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                openAutoStartSettings();
            }
        });
    }

    private void openAutoStartSettings() {
        boolean opened = false;
        String manufacturer = Build.MANUFACTURER.toLowerCase();
        LogCollector.addLog("AUTOSTART", "Manufacturer: " + manufacturer);

        try {
            Intent intent = new Intent();
            if (manufacturer.contains("xiaomi") || manufacturer.contains("redmi")) {
                intent.setComponent(new ComponentName("com.miui.securitycenter",
                        "com.miui.permcenter.autostart.AutoStartManagementActivity"));
                startActivity(intent);
                opened = true;
            } else if (manufacturer.contains("huawei") || manufacturer.contains("honor")) {
                intent.setComponent(new ComponentName("com.huawei.systemmanager",
                        "com.huawei.systemmanager.startupmgr.ui.StartupNormalAppListActivity"));
                startActivity(intent);
                opened = true;
            } else if (manufacturer.contains("oppo")) {
                intent.setComponent(new ComponentName("com.coloros.safecenter",
                        "com.coloros.safecenter.permission.startup.StartupAppListActivity"));
                startActivity(intent);
                opened = true;
            } else if (manufacturer.contains("vivo")) {
                intent.setComponent(new ComponentName("com.vivo.permissionmanager",
                        "com.vivo.permissionmanager.activity.BgStartUpManagerActivity"));
                startActivity(intent);
                opened = true;
            } else if (manufacturer.contains("realme")) {
                intent.setComponent(new ComponentName("com.coloros.safecenter",
                        "com.coloros.safecenter.startupapp.StartupAppListActivity"));
                startActivity(intent);
                opened = true;
            } else if (manufacturer.contains("oneplus")) {
                intent.setComponent(new ComponentName("com.oneplus.security",
                        "com.oneplus.security.chainlaunch.view.ChainLaunchAppListActivity"));
                startActivity(intent);
                opened = true;
            } else if (manufacturer.contains("samsung")) {
                intent.setComponent(new ComponentName("com.samsung.android.lool",
                        "com.samsung.android.sm.ui.battery.BatteryActivity"));
                startActivity(intent);
                opened = true;
            } else if (manufacturer.contains("asus")) {
                intent.setComponent(new ComponentName("com.asus.mobilemanager",
                        "com.asus.mobilemanager.autostart.AutoStartActivity"));
                startActivity(intent);
                opened = true;
            } else if (manufacturer.contains("lenovo")) {
                intent.setComponent(new ComponentName("com.lenovo.security",
                        "com.lenovo.security.purebackground.PureBackgroundActivity"));
                startActivity(intent);
                opened = true;
            } else if (manufacturer.contains("tecno") || manufacturer.contains("infinix") || manufacturer.contains("itel")) {
                intent.setComponent(new ComponentName("com.transsion.phonemanager",
                        "com.transsion.phonemanager.module.autostart.AutoStartActivity"));
                startActivity(intent);
                opened = true;
            }
        } catch (Exception e) {
            opened = false;
            LogCollector.addWarn("AUTOSTART", "Primary intent failed: " + e.getMessage());
        }

        if (!opened) {
            try {
                Intent intent = new Intent();
                intent.setComponent(new ComponentName("com.miui.securitycenter",
                        "com.miui.permcenter.autostart.AutoStartManagementActivity"));
                startActivity(intent);
                opened = true;
            } catch (Exception e) {}
        }

        if (!opened) {
            try {
                Intent intent = new Intent();
                intent.setComponent(new ComponentName("com.huawei.systemmanager",
                        "com.huawei.systemmanager.optimize.process.ProtectActivity"));
                startActivity(intent);
                opened = true;
            } catch (Exception e) {}
        }

        if (!opened) {
            try {
                Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
                intent.setData(Uri.parse("package:" + getPackageName()));
                startActivity(intent);
                opened = true;
                Toast.makeText(this, "Please enable AutoStart/Background permission manually", Toast.LENGTH_LONG).show();
            } catch (Exception e) {
                LogCollector.addError("AUTOSTART", "Cannot open any settings", e);
            }
        }

        if (opened) {
            LogCollector.addLog("AUTOSTART", "Opened autostart settings for " + manufacturer);
            Toast.makeText(this, "Please enable AutoStart for Cherry SME TTS", Toast.LENGTH_LONG).show();
        }
    }

    private void updateBatteryStatus() {
        try {
            PowerManager pm = (PowerManager) getSystemService(Context.POWER_SERVICE);
            TextView btnBattery = findViewById(R.id.btnBatteryOptimize);
            if (pm != null && pm.isIgnoringBatteryOptimizations(getPackageName())) {
                btnBattery.setText("Battery Optimization Disabled");
                btnBattery.setTextColor(0xFF388E3C);
            } else {
                btnBattery.setText("Disable Battery Optimization");
                btnBattery.setTextColor(0xFFE65100);
            }
        } catch (Exception e) {}
    }

    @Override
    protected void onResume() {
        super.onResume();
        updateBatteryStatus();
    }

    private void showScanDialog() {
        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);

        progressDialog = new ProgressDialog(this);
        progressDialog.setMessage("Scanning engines, please wait...");
        progressDialog.setCancelable(false);
        progressDialog.show();
    }

    private void setupDonation(int viewId, final String number, final String msg) {
        View btn = findViewById(viewId);
        btn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                ClipboardManager clipboard = (ClipboardManager) getSystemService(Context.CLIPBOARD_SERVICE);
                ClipData clip = ClipData.newPlainText("Donation Number", number);
                clipboard.setPrimaryClip(clip);
                Toast.makeText(MainActivity.this, msg, Toast.LENGTH_SHORT).show();
            }
        });

        if (viewId == R.id.btnKpay) {
            btn.setOnLongClickListener(new View.OnLongClickListener() {
                @Override
                public boolean onLongClick(View v) {
                    try {
                        startActivity(new Intent(MainActivity.this, LogViewerActivity.class));
                    } catch (Exception e) {
                        Toast.makeText(MainActivity.this, "Error opening logs", Toast.LENGTH_SHORT).show();
                    }
                    return true;
                }
            });
        }
    }

    private void setupEngineUI(int spinnerId, String pkgKey, String defPkg) {
        Spinner spinner = findViewById(spinnerId);
        setSpinnerSelection(spinner, pkgKey, defPkg);
    }

    private void setupOpenSystemSettings(int viewId) {
        View view = findViewById(viewId);
        view.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                try {
                    Intent intent = new Intent();
                    intent.setAction("com.android.settings.TTS_SETTINGS");
                    intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                    startActivity(intent);
                    Toast.makeText(MainActivity.this, "Opening System TTS Settings...", Toast.LENGTH_SHORT).show();
                } catch (Exception e) {
                    Toast.makeText(MainActivity.this, "Cannot open settings", Toast.LENGTH_SHORT).show();
                }
                return true;
            }
        });
    }

    private void loadInstalledEngines() {
        Intent intent = new Intent("android.intent.action.TTS_SERVICE");
        List<ResolveInfo> resolveInfos = getPackageManager().queryIntentServices(intent, PackageManager.GET_META_DATA);
        engineNames.clear();
        enginePackages.clear();

        for (ResolveInfo info : resolveInfos) {
            String pkg = info.serviceInfo.packageName;
            if (!pkg.equals(getPackageName())) {
                String label = info.serviceInfo.loadLabel(getPackageManager()).toString();
                engineNames.add(label);
                enginePackages.add(pkg);
            }
        }

        if (engineNames.isEmpty()) {
            engineNames.add("No Engines Found");
            enginePackages.add("");
        }

        ArrayAdapter<String> adapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_dropdown_item, engineNames);

        Spinner spShan = findViewById(R.id.spinnerShan);
        Spinner spBur = findViewById(R.id.spinnerBurmese);
        Spinner spEng = findViewById(R.id.spinnerEnglish);

        spShan.setAdapter(adapter);
        spBur.setAdapter(adapter);
        spEng.setAdapter(adapter);
    }

    private void setSpinnerSelection(Spinner spinner, final String key, String def) {
        String saved = prefs.getString(key, def);
        int idx = enginePackages.indexOf(saved);
        if (idx >= 0) spinner.setSelection(idx);

        spinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                if (!enginePackages.isEmpty() && position >= 0) {
                    prefs.edit().putString(key, enginePackages.get(position)).apply();
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {}
        });
    }
}

File: app/src/main/java/com/shan/tts/manager/MainActivity.kt
package com.shan.tts.manager

import android.app.ProgressDialog
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.os.Bundle
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity

class MainActivity : AppCompatActivity() {

    private lateinit var prefs: SharedPreferences
    private lateinit var configPrefs: SharedPreferences
    private val engineNames = ArrayList<String>()
    private val enginePackages = ArrayList<String>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        prefs = getSharedPreferences("TTS_SETTINGS", Context.MODE_PRIVATE)
        configPrefs = getSharedPreferences("TTS_CONFIG", Context.MODE_PRIVATE)

        loadInstalledEngines()

        setupEngineUI(R.id.spinnerShan, "pref_shan_pkg", "com.espeak.ng", 
            R.id.seekShanRate, "SHAN_RATE", 
            R.id.seekShanPitch, "SHAN_PITCH")
            
        setupEngineUI(R.id.spinnerBurmese, "pref_burmese_pkg", "com.google.android.tts", 
            R.id.seekBurmeseRate, "MYANMAR_RATE", 
            R.id.seekBurmesePitch, "MYANMAR_PITCH")
            
        setupEngineUI(R.id.spinnerEnglish, "pref_english_pkg", "com.google.android.tts", 
            R.id.seekEnglishRate, "ENGLISH_RATE", 
            R.id.seekEnglishPitch, "ENGLISH_PITCH")

        setupDonation(R.id.btnKpay, "09750091817", "KBZ Pay Number Copied")
        setupDonation(R.id.btnWave, "09750091817", "Wave Pay Number Copied")

        performFullSystemScan()
    }

    private fun performFullSystemScan() {
        val progress = ProgressDialog(this)
        progress.setTitle("System Preparation")
        progress.setMessage("Initializing all TTS Engines...\nPlease wait, this may take a while.")
        progress.setCancelable(false) 
        progress.show()

        EngineScanner.scanAllEngines(this) {
            runOnUiThread {
                try {
                    if (progress.isShowing) {
                        progress.dismiss()
                    }
                } catch (e: Exception) {}
                Toast.makeText(this, "All Engines Ready!", Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun setupDonation(viewId: Int, number: String, msg: String) {
        val btn = findViewById<View>(viewId)
        btn?.setOnClickListener {
            val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
            val clip = ClipData.newPlainText("Donation Number", number)
            clipboard.setPrimaryClip(clip)
            Toast.makeText(this, msg, Toast.LENGTH_SHORT).show()
        }
        if (viewId == R.id.btnKpay) {
            btn?.setOnLongClickListener {
                try {
                    startActivity(Intent(this, LogViewerActivity::class.java))
                } catch (e: Exception) {}
                true
            }
        }
    }

    private fun setupEngineUI(spinnerId: Int, pkgKey: String, defPkg: String, rateId: Int, rateKey: String, pitchId: Int, pitchKey: String) {
        val spinner = findViewById<Spinner>(spinnerId)
        setSpinnerSelection(spinner, pkgKey, defPkg)

        spinner.setOnLongClickListener {
            try {
                val intent = Intent()
                intent.action = "com.android.settings.TTS_SETTINGS"
                intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
                startActivity(intent)
                Toast.makeText(this, "Opening System Settings...", Toast.LENGTH_SHORT).show()
            } catch (e: Exception) {
                Toast.makeText(this, "Cannot open settings directly", Toast.LENGTH_SHORT).show()
            }
            true
        }

        val seekRate = findViewById<SeekBar>(rateId)
        val seekPitch = findViewById<SeekBar>(pitchId)

        seekRate.max = 200 
        seekPitch.max = 200 

        seekRate.progress = configPrefs.getInt(rateKey, 100)
        seekPitch.progress = configPrefs.getInt(pitchKey, 100)

        seekRate.setOnSeekBarChangeListener(getSeekListener(rateKey))
        seekPitch.setOnSeekBarChangeListener(getSeekListener(pitchKey))
    }

    private fun getSeekListener(key: String): SeekBar.OnSeekBarChangeListener {
        return object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar?, progress: Int, fromUser: Boolean) {
                val safeProgress = if (progress < 10) 10 else progress
                if (fromUser) configPrefs.edit().putInt(key, safeProgress).apply()
            }
            override fun onStartTrackingTouch(seekBar: SeekBar?) {}
            override fun onStopTrackingTouch(seekBar: SeekBar?) {}
        }
    }

    private fun loadInstalledEngines() {
        val intent = Intent("android.intent.action.TTS_SERVICE")
        val resolveInfos = packageManager.queryIntentServices(intent, PackageManager.GET_META_DATA)
        engineNames.clear(); enginePackages.clear()

        for (info in resolveInfos) {
            val pkg = info.serviceInfo.packageName
            if (pkg != packageName) {
                val label = info.serviceInfo.loadLabel(packageManager).toString()
                engineNames.add(label)
                enginePackages.add(pkg)
            }
        }
        
        if (engineNames.isEmpty()) { 
            engineNames.add("No Engines Found")
            enginePackages.add("") 
        }

        val adapter = ArrayAdapter(this, android.R.layout.simple_spinner_dropdown_item, engineNames)
        
        val spShan = findViewById<Spinner>(R.id.spinnerShan)
        val spBur = findViewById<Spinner>(R.id.spinnerBurmese)
        val spEng = findViewById<Spinner>(R.id.spinnerEnglish)

        spShan?.adapter = adapter
        spBur?.adapter = adapter
        spEng?.adapter = adapter
    }

    private fun setSpinnerSelection(spinner: Spinner?, key: String, def: String) {
        val saved = prefs.getString(key, def)
        val idx = enginePackages.indexOf(saved)
        if (idx >= 0) spinner?.setSelection(idx)
        
        spinner?.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(p: AdapterView<*>?, v: View?, pos: Int, id: Long) {
                if (enginePackages.isNotEmpty() && pos >= 0) {
                    prefs.edit().putString(key, enginePackages[pos]).apply()
                }
            }
            override fun onNothingSelected(p: AdapterView<*>?) {}
        }
    }
}

